server-timing/class-perflab-server-timing.php

```diff
@@ -0,0 +1,265 @@
+
+	/**
+	 * Registers a metric to calculate for the Server-Timing header.
+	 *
+	 * This method must be called before the {@see 'perflab_server_timing_send_header'} hook.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $metric_slug The metric slug.
+	 * @param array  $args        {
+	 *     Arguments for the metric.
+	 *
+	 *     @type callable $measure_callback The callback that initiates calculating the metric value. It will receive
+	 *                                      the Perflab_Server_Timing_Metric instance as a parameter, in order to set
+	 *                                      the value when it has been calculated. Metric values must be provided in
+	 *                                      milliseconds.
+	 *     @type string   $access_cap       Capability required to view the metric. If this is a public metric, this
+	 *                                      needs to be set to "exist".
+	 * }
+	 */
+	public function register_metric( $metric_slug, array $args ) {
+		if ( isset( $this->registered_metrics[ $metric_slug ] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: metric slug */
+				sprintf( __( 'A metric with the slug %s is already registered.', 'performance-lab' ), $metric_slug ),
+				''
+			);
+			return;
+		}
+
+		if ( did_action( 'perflab_server_timing_send_header' ) && ! doing_action( 'perflab_server_timing_send_header' ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: WordPress action name */
+				sprintf( __( 'The method must be called before or during the %s action.', 'performance-lab' ), 'perflab_server_timing_send_header' ),
+				''
+			);
+			return;
+		}
+
+		$args = wp_parse_args(
+			$args,
+			array(
+				'measure_callback' => null,
+				'access_cap'       => null,
+			)
+		);
+		if ( ! $args['measure_callback'] || ! is_callable( $args['measure_callback'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: PHP parameter name */
+				sprintf( __( 'The %s argument is required and must be a callable.', 'performance-lab' ), '$args["measure_callback"]' ),
+				''
+			);
+			return;
+		}
+		if ( ! $args['access_cap'] || ! is_string( $args['access_cap'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: PHP parameter name */
+				sprintf( __( 'The %s argument is required and must be a string.', 'performance-lab' ), '$args["access_cap"]' ),
+				''
+			);
+			return;
+		}
+
+		$this->registered_metrics[ $metric_slug ]      = new Perflab_Server_Timing_Metric( $metric_slug );
+		$this->registered_metrics_data[ $metric_slug ] = $args;
+
+		// If the current user has already been determined and they lack the necessary access,
+		// do not even attempt to calculate the metric.
+		if ( did_action( 'set_current_user' ) && ! current_user_can( $args['access_cap'] ) ) {
+			return;
+		}
+
+		// Otherwise, call the measuring callback and pass the metric instance to it.
+		call_user_func( $args['measure_callback'], $this->registered_metrics[ $metric_slug ] );
```

Comment:
I'm curious why this method doesn't rather `return` the `$this->registered_metrics[ $metric_slug ]` value here? Why is `measure_callback` needed?

Then instead of:

https://github.com/WordPress/performance/blob/0e10d5a0a924bd1d35345587eabb2984814c502d/server-timing/load.php#L105-L115

It could be simplified to:

```php
		if ( ! perflab_server_timing()->has_registered_metric( $metric_slug ) ) {
			$server_timing_metric = perflab_server_timing_register_metric(
				$metric_slug,
				array(
					'access_cap' => $access_cap,
				)
			);
		}
```

Reply:
@westonruter I think that scenario is an edge-case, which is only needed because of the way this function is a wrapper to an action callback.

Overall, I consider that a discouraged pattern. Passing a measurement callback allows for that callback to only be called on other conditions (e.g. depending on the "access cap"). If the metric was returned, it would encourage implementing the measurement logic there, which then would be decoupled from those conditions for whether the metric should be measured in the first place. Alternatively, we could return the metric only if the conditions are satisfied, but that would be a strange pattern IMO, as not returning it wouldn't mean that the metric wasn't registered correctly - whereas the return value of a "register" function would be expected to only return `null` or `false` if the registration itself failed.